tests/models/musicgen/test_modeling_musicgen.py

```diff
@@ -498,6 +500,89 @@ def prepare_config_and_inputs_for_common(self):
+    """
+
+    def __init__(self, model: "MusicgenForConditionalGeneration", play_steps: int = 10):
+        self.decoder = model.decoder
+        self.audio_encoder = model.audio_encoder
+        self.generation_config = model.generation_config
+
+        # variables used in the streaming process
+        self.play_steps = play_steps
+        self.token_cache = None
+        self.to_print = 0
+
+    def apply_delay_pattern_mask(self, input_ids):
+        # build the delay pattern mask for offsetting each codebook prediction by 1 (this behaviour is specific to MusicGen)
+        _, decoder_delay_pattern_mask = self.decoder.build_delay_pattern_mask(
+            self.token_cache[:, :1],
+            pad_token_id=self.generation_config.decoder_start_token_id,
+            max_length=self.token_cache.shape[-1],
+        )
+        # apply the pattern mask to the input ids
+        input_ids = self.decoder.apply_delay_pattern_mask(input_ids, decoder_delay_pattern_mask)
+
+        # revert the pattern delay mask by filtering the pad token id
+        input_ids = input_ids[input_ids != self.generation_config.pad_token_id].reshape(
+            1, self.decoder.num_codebooks, -1
+        )
+
+        # append the frame dimension back to the audio codes
+        input_ids = input_ids[None, ...]
+
+        # send the input_ids to the correct device
+        input_ids = input_ids.to(self.audio_encoder.device)
+
+        output_values = self.audio_encoder.decode(
+            input_ids,
+            audio_scales=[None],
+        )
+        audio_values = output_values.audio_values[0, 0]
+        return audio_values.cpu().numpy()
+
+    def put(self, value):
+        batch_size = value.shape[0] // self.decoder.num_codebooks
+        if batch_size > 1:
+            raise ValueError("MusicgenStreamer only supports batch size 1")
+
+        if self.token_cache is None:
+            self.token_cache = value
+        else:
+            self.token_cache = torch.concatenate([self.token_cache, value[:, None]], dim=-1)
+
+        if self.token_cache.shape[-1] % self.play_steps == 0:
+            audio_values = self.apply_delay_pattern_mask(self.token_cache)
+            # only print the newly generated audio values
+            audio_values = audio_values[self.to_print :]
+            self.on_finalized_text(audio_values)
+            # update our running total
+            self.to_print += len(audio_values)
+
+    def end(self):
+        """Flushes any remaining cache and prints a newline to stdout."""
+        # Flush the cache, if it exists
+        if self.token_cache is not None:
+            audio_values = self.apply_delay_pattern_mask(self.token_cache)
+        else:
+            audio_values = ""
+
+        self.on_finalized_text(audio_values[self.to_print :], stream_end=True)
+        self.to_print += len(audio_values)
+
+    def on_finalized_text(self, audio: np.ndarray, stream_end: bool = False):
```

Comment:
Nit: Since this is an internal function, perhaps `on_finalized_chunk` (or another non-text name) :)

Reply:
Resolved in https://github.com/huggingface/transformers/pull/25320/commits/58de4384a1a33f6b69754b3026950bc4a5534992